### PR TITLE
feat: Add documentation about S3 Object Lock

### PR DESCRIPTION
**Pull request by revij**
Explain how one can use object locking in S3 to prevent data deletion or force retention periods.

### CODE REVIEW

**Comment by fghaas** [comment]
By the way: the test failures are trivial. Just run `tox -e gitlint` and `tox` locally and the output should be self-explanatory. :)

File: docs/howto/object-storage/s3/object-lock.md [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+enabled/disabled for any object in a locked bucket, regardless of the 
+lock configuration, object retention, or age.
+
+This feature provides the same level of protection as a retention period 
+but with no fixed expiration date. Instead, a legal hold will remain in 
+effect until you remove it explicitly.
+
+## Prerequisites
+
+- [Aws cli installed](index.md)
```
[the remark this answers is not in the document]
Comment (revij) [reply docs/howto/object-storage/s3/object-lock.md:75]: Indeed this is limited to the `aws` cli tool as it was the easiest approach.
I will see how we can better include `mc` (you need to specify a json file with `mc ilm import ` - https://min.io/docs/minio/linux/reference/deprecated/mc-ilm-import.html) and `s3cmd` (by using the `setlifecycle` option along with a xml file)

File: docs/howto/object-storage/s3/object-lock.md [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+In order to bypass the governance mode, you must indicate the
+`--bypass-governance-retention` flag when deleting the object:
+
+```bash
+aws --profile <region> \
+    --endpoint-url=https://s3-<region>.citycloud.com:8080 \
+    s3api delete-object \
+    --bucket <bucket-name> \
+    --key <object-name> \
+    --bypass-governance-retention
```
[the remark this answers is not in the document]
Comment (revij) [reply docs/howto/object-storage/s3/object-lock.md:179]: Indeed, including `--bypass-governance-retention` will automatically include `'x-amz-bypass-governance-retention': b'True'` on the headers list (tested and confirmed working).

**Comment by revij** [comment]
Pushed a few more commits now to include `mc`, syntax edits and split up some parts. I’m a bit unsure as to what is the best approach for the “bucket-level” / “Per bucket” headings since a markdown validation will fail if both are named the same (no-duplicate-heading)

**Comment by fghaas** [comment]
@revij Just FYI, within a PR review we're perfectly cool with you continuing to amend your commit with `git commit -a --amend` and updating your topic branch with `git push --force-with-lease`.

You can squash your commits into one now, too. Or else I can do it for you. Let me know what you prefer.

File: docs/howto/object-storage/s3/object-lock.md [hunk 3]
```diff
@@ -0,0 +1,361 @@
+# Object Lock
```
Comment (fghaas) [inline docs/howto/object-storage/s3/object-lock.md:1]: To be in line with the other page headings, "Object versioning" and "Object expiry", please lowercase the word "lock":

```suggestion
# Object lock
```

Do the same for other mentions of that term too, please.

File: docs/howto/object-storage/s3/object-lock.md [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+> allowing only authorized users to perform any
+> necessary actions while effectively managing objects.
+
+Governance mode is ideal for scenarios where you want to prevent
+most users from deleting objects during a specific retention period,
+but still allow authorized users with special permissions to modify
+the retention settings or delete objects as needed.
+
+To override or remove governance-mode retention settings, users must
+have the `s3:BypassGovernanceRetention` permission. With this permission,
```
Comment (fghaas) [inline docs/howto/object-storage/s3/object-lock.md:42]: In our environment, how do users gain this permission? How can they find out whether their user account has it?

File: docs/howto/object-storage/s3/object-lock.md [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+        --endpoint-url=https://s3-<region>.citycloud.com:8080 \
+        s3api put-object-retention \
+        --bucket <bucket-name> \
+        --key <object-name> \
+        --retention \
+        '{ "Mode": "GOVERNANCE", "RetainUntilDate": "2023-12-31T00:00:00Z" }'
+    ```
+    > For the value of the `RetainUntilDate` parameter, use the
+    > [ISO 8601 date-time representation](https://en.wikipedia.org/wiki/ISO_8601#Combined_date_and_time_representations)
+    > (`Y-m-dTH:M:S.%3fZ`) format.
```
Comment (fghaas) [inline docs/howto/object-storage/s3/object-lock.md:125]: I think the `Y-m-dTH:M:S.%3fZ` format string is more confusing than beneficial. :) Consider dropping it.

File: docs/howto/object-storage/s3/object-lock.md [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+    --default governance "2d"
+    ```
+    > `--default` sets the default object lock settings for new objects and
+    > is optional.
+
+    > To specify a duration, use a string formatted as `Nd`
+    > for days or `Ny` for years.
+    > For example, use `30d` to indicate 30 days after
+    > the object creation, or use `1y` to indicate 1 year
+    > after the object creation.
```
Comment (fghaas) [inline docs/howto/object-storage/s3/object-lock.md:139]: How about `s3cmd` (here, and in the remainder of the command examples)?

**Comment by fghaas** [comment]
> You can squash your commits into one now, too. Or else I can do it for you. Let me know what you prefer.

@revij Can you let me know your preference please?

**Comment by revij** [comment]
> > You can squash your commits into one now, too. Or else I can do it for you. Let me know what you prefer.
> 
> @revij Can you let me know your preference please?

Should be ready now!

File: docs/howto/object-storage/s3/object-lock.md (hunk 4 above)
Comment (revij) [reply docs/howto/object-storage/s3/object-lock.md]: I've removed the governance mode mention as we need to work on that part.

File: docs/howto/object-storage/s3/object-lock.md [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+Retention periods specify the length of time that an object
+remains locked and protected from being overwritten or deleted.
+When you set a retention period, the object is safeguarded for
+the specified time period. You can set the retention period in
+either days or years, with a **minimum of one day** and
+no maximum limit.
+
+In addition to retention periods, you can choose the retention
+mode that applies to your objects. At the moment only the
+**Compliance** mode is available.
```
Comment (fghaas) [inline docs/howto/object-storage/s3/object-lock.md:28]: I prefer "currently" over "at the moment", because the former is neutral and the latter hints that something is about to change soon.

Also, if of the two object lock retention modes we support only one, we need an entry to that effect in the object storage of the Limitations reference (`docs/reference/limitations/object-storage.md`).

File: docs/howto/object-storage/s3/object-lock.md [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+as it prevents objects from being overwritten or deleted
+by any user. If you configure an object with this mode, you
+cannot shorten or change its retention period, ensuring that
+the data remains **secure** and **compliant** with regulatory
+requirements.
+
+> Should be used when you want to ensure that no user, including
+> the administrator user, can delete objects during a defined
+> retention period.
+
```
Comment (fghaas) [inline docs/howto/object-storage/s3/object-lock.md:45]: This section was previously a side-by-side comparison between Governance and Compliance mode, where it made sense. Now that the bit about Governance mode has been thrown out, it feels oddly disconnected.

Can we edit it down to this?

```suggestion
Compliance mode is recommended when storing compliant data,
as it prevents objects from being overwritten or deleted
by any user. If you configure an object with this mode, you
cannot shorten or change its retention period, ensuring that
the data remains **secure** and **compliant** with regulatory
requirements.
```

File: docs/howto/object-storage/s3/object-lock.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (fghaas) [reply docs/howto/object-storage/s3/object-lock.md:59]: Right, but now the link text no longer makes sense.

File: docs/howto/object-storage/s3/object-lock.md [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+lock configuration, object retention, or age.
+
+This feature provides the same level of protection as a retention period
+but with no fixed expiration date. Instead, a legal hold will remain in
+effect until you remove it explicitly.
+
+## Prerequisites
+
+- [Aws cli installed](index.md)
+- [S3-compatible credentials](credentials.md)
```
Comment (fghaas) [inline docs/howto/object-storage/s3/object-lock.md:60]: Let's be specific and explicit about the fact that we recommend against using `s3cmd` for this feature:
```suggestion
In order to manage object lock, you must install `aws` or `mc`, and [configure it with working credentials](credentials.md).

You cannot (reliably) use `s3cmd` to manage this functionality.
```

File: docs/howto/object-storage/s3/object-lock.md [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+    aws --profile <region> \
+      --endpoint-url=https://s3-<region>.citycloud.com:8080 \
+      s3api create-bucket \
+      --bucket <bucket-name> \
+      --object-lock-enabled-for-bucket
+    ```
+=== "mc"
+    ```bash
+    mc mb <region>/<bucket> \
+    --with-lock
```
Comment (fghaas) [inline docs/howto/object-storage/s3/object-lock.md:81]: This is missing an additional two-space indent on the second line:
```suggestion
      --with-lock
```

File: docs/howto/object-storage/s3/object-lock.md [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+        '{ "Mode": "COMPLIANCE", "RetainUntilDate": "2023-12-31T00:00:00Z" }'
+    ```
+    > For the value of the `RetainUntilDate` parameter, use the
+    > [ISO 8601 date-time representation](https://en.wikipedia.org/wiki/ISO_8601#Combined_date_and_time_representations)
+    > format.
+=== "mc"
+    ```bash
+    mc retention set \
+    <region>/<bucket> \
+    --default compliance "2d"
```
Comment (fghaas) [inline docs/howto/object-storage/s3/object-lock.md:108]:
```suggestion
      <region>/<bucket> \
      --default compliance "2d"
```

File: docs/howto/object-storage/s3/object-lock.md [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+        s3api put-legal-hold \
+        --bucket <bucket-name> \
+        --key <object-name> \
+        --legal-hold Status=ON
+    ```
+=== "mc"
+    ```bash
+    mc legalhold set \
+    --recursive \
+    <region>/<bucket>
```
Comment (fghaas) [inline docs/howto/object-storage/s3/object-lock.md:229]:
```suggestion
      --recursive \
      <region>/<bucket>
```

File: docs/howto/object-storage/s3/object-lock.md [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+        --endpoint-url=https://s3-<region>.citycloud.com:8080 \
+        s3api put-legal-hold \
+        --bucket <bucket-name> \
+        --key <object-name> \
+        --legal-hold Status=ON
+    ```
+=== "mc"
+    ```bash
+    mc legalhold set \
+    <region>/<bucket>/<object-name>
```
Comment (fghaas) [inline docs/howto/object-storage/s3/object-lock.md:249]:
```suggestion
      <region>/<bucket>/<object-name>
```

File: docs/howto/object-storage/s3/object-lock.md [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+    {
+        "LegalHold": {
+            "Status": "ON"
+        }
+    }
+    ```
+=== "mc"
+    ```bash
+    mc legalhold info \
+    --json <region>/<bucket>/<object-name>
```
Comment (fghaas) [inline docs/howto/object-storage/s3/object-lock.md:273]:
```suggestion
      --json <region>/<bucket>/<object-name>
```

File: docs/howto/object-storage/s3/object-lock.md [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+To use the object lock feature, you must first create a new bucket.
+
+> Existing buckets cannot have object lock enabled.
+> If you require assistance to enable object lock on an existing bucket,
+> please contact our support team.
+
+=== "aws"
+    ```bash
+    aws --profile <region> \
+      --endpoint-url=https://s3-<region>.citycloud.com:8080 \
```
Comment (fghaas) [inline docs/howto/object-storage/s3/object-lock.md:65]: The `--endpoint-url` option is no longer necessary to add here (we are now instructing users to use a plugin for managing endpoints, since #139).

File: docs/howto/object-storage/s3/object-lock.md (hunk 7 above)
Comment (revij) [reply docs/howto/object-storage/s3/object-lock.md]: Saw that you already took care of this, thanks :)

**Comment by fghaas** [comment]
All the remaining minor nits are now fixed, and the branch is rebased on current `main`. @colder-is-better, at your convenience I would appreciate an extra pair of eyes just to make sure I didn't miss anything.

File: docs/howto/object-storage/s3/object-lock.md [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+> objects? Then you should check our [Object expiry](expiry.md) guide.
+
+## Object lock modes
+
+Object lock offers two methods for managing object retention:
+**Retention periods** and **Legal hold**.
+
+### Retention periods
+
+Retention periods specify the length of time that an object
```
Comment (colder-is-better) [inline docs/howto/object-storage/s3/object-lock.md:19]:
```suggestion
Retention periods specify the length of time an object
```

File: docs/howto/object-storage/s3/object-lock.md [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+
+You cannot (reliably) use `s3cmd` to manage this functionality.
+
+## Enabling object lock
+
+To use the object lock feature, you must first create a new bucket.
+
+> Existing buckets cannot have object lock enabled.
+> If you require assistance to enable object lock on an existing bucket,
+> please contact our support team.
```
Comment (colder-is-better) [inline docs/howto/object-storage/s3/object-lock.md:60]:
```suggestion
> please get in touch with our support team.
```

**Review by colder-is-better — COMMENTED** [review]
It's practically flawless (I only have two subatomic nits).

File: docs/howto/object-storage/s3/object-lock.md (hunk 17 above)
Comment (fghaas) [reply docs/howto/object-storage/s3/object-lock.md:60]: Actually, on that note. It just now occurs to me that that paragraph is kind of a contradiction, because if "existing buckets *cannot* have object lock enabled" (emphasis mine), then it doesn't make sense to contact support either.

So we either need to rephrase this to reflect that *customers* cannot enable object lock on an existing bucket (but we can), or else drop the support reference altogether.

@revij, can you clarify please?

Comment (revij) [reply docs/howto/object-storage/s3/object-lock.md:60]: The idea here is that if a customer has a large bucket and wants to have object locking enabled, we can create that bucket ourselves and copy over the data, otherwise the customer will have to pay 2x the ingress fee to create a new bucket and upload everything again.

Comment (fghaas) [reply docs/howto/object-storage/s3/object-lock.md]: No worries, fixed in [97463f3](https://github.com/citynetwork/docs/pull/135/commits/97463f3880f90bf0587ad81e2e0ae94cd5517c67) :)

File: docs/howto/object-storage/s3/object-lock.md (hunk 16 above)
Comment (fghaas) [reply docs/howto/object-storage/s3/object-lock.md:19]: I prefer to leave "that" in, in this instance.

**Comment by fghaas** [comment]
As a (hopefully) final change prior to merging, I have added a description (in the YAML front matter) and condensed the opening paragraph I tiny bit.

**Review by fghaas — APPROVED** [review]
As far as I am concerned, this is good to merge now. If anyone has objections, please speak up promptly. :)
